core-plugins/src/main/java/io/cdap/plugin/transform/JavaScriptTransform.java

```diff
@@ -346,45 +356,51 @@ private Object decodeSimpleType(Object object, Schema schema) {
       case STRING:
         return (String) object;
     }
-    throw new RuntimeException("Unable decode object with schema " + schema);
+    collector.addFailure(
+        String.format("Unable to decode object with schema %s", schema),
+        "Ensure the object can be decoded with the input schema").withConfigProperty("script");
+    throw collector.getOrThrowException();
   }
 
-  private Map<Object, Object> decodeMap(Map<Object, Object> object, Schema keySchema, Schema valSchema) {
+  private Map<Object, Object> decodeMap(Map<Object, Object> object, Schema keySchema, Schema valSchema,
+      FailureCollector collector) {
     Map<Object, Object> output = Maps.newHashMap();
     for (Map.Entry<Object, Object> entry : object.entrySet()) {
-      output.put(decode(entry.getKey(), keySchema), decode(entry.getValue(), valSchema));
+      output.put(decode(entry.getKey(), keySchema, collector), decode(entry.getValue(), valSchema, collector));
     }
     return output;
   }
 
-  private List<Object> decodeArray(List nativeArray, Schema componentSchema) {
+  private List<Object> decodeArray(List nativeArray, Schema componentSchema, FailureCollector collector) {
     List<Object> arr = Lists.newArrayListWithCapacity(nativeArray.size());
     for (Object arrObj : nativeArray) {
-      arr.add(decode(arrObj, componentSchema));
+      arr.add(decode(arrObj, componentSchema, collector));
     }
     return arr;
   }
 
-  private Object decodeUnion(Object object, List<Schema> schemas) {
+  private Object decodeUnion(Object object, List<Schema> schemas, FailureCollector collector) {
     for (Schema schema : schemas) {
       try {
-        return decode(object, schema);
+        return decode(object, schema, collector);
       } catch (Exception e) {
         // could be ok, just move on and try the next schema
       }
     }
-
-    throw new RuntimeException("Unable decode union with schema " + schemas);
+    collector.addFailure(String.format("Unable to decode union with schema %s", schemas),
+        "Ensure the provided schema can be used to decode the union").withConfigProperty("script");
+    throw collector.getOrThrowException();
   }
 
-  private void init(@Nullable TransformContext context) {
+  private void init(@Nullable TransformContext context, FailureCollector collector) {
     ScriptEngineManager manager = new ScriptEngineManager();
     engine = manager.getEngineByName("JavaScript");
     try {
       engine.eval(ScriptConstants.HELPER_DEFINITION);
     } catch (ScriptException e) {
       // shouldn't happen
-      throw new IllegalStateException("Couldn't define helper functions", e);
+      collector.addFailure("Couldn't define helper functions", "Something went wrong");
```

Comment:
keep corrective action null in this case.

Reply:
period at the end of error message

Reply:
Fixed